### PR TITLE
Cleaner expressions

### PR DESCRIPTION
**Pull request by gwicke**
Implementation of https://phabricator.wikimedia.org/T118700.

### CODE REVIEW

**Comment by gwicke** [comment]
Confirmed working with https://github.com/wikimedia/restbase/pull/415, which verifies that this code preserves backwards compatibility.

**Comment by gwicke** [comment]
See also:  wikimedia/restbase#417.

File: lib/reqTemplate.js [hunk 1]
```diff
@@ -69,35 +106,46 @@ function splitAndPrepareTAsseblyTemplate(templateSpec, options) {
...
             if (templateNest === 0) { // We are either entering a new template
                 if (startIndex !== index) {
                     result.push(templateSpec.substring(startIndex, index));
                 }
+                if (templateSpec[index + 1] === '{') {
+                    index++;
+                    inDoubleBrace = true;
+                }
                 startIndex = index + 1;
+
```
Comment (d00rman) [inline lib/reqTemplate.js]: Why all these empty lines in the middle?

**Comment by d00rman** [comment]
Thinking now about double braces, wouldn't it be better to simply have them in URIs alone? I mena, escaping in other parts of the request makes little to no sense, while on the other hand it requires the user to switch context between request parts, which I think is not that desirable. Compare these two:

``` yaml
reqA:
  uri: {{host}}/{path}
  headers:
    x-first: {request.foo}
  body:
    field: {request.bar}

reqB:
  uri: {{host}}/{path}
  headers:
    x-first: {{request.foo}}
  body:
    field: {{request.bar}}
```

It feels more natural to escape parts of the URI then **not to** escape the headers or body.

File: lib/reqTemplate.js (hunk 1 above)
Comment (gwicke) [inline lib/reqTemplate.js]: Removed this one in a follow-up. Others below are intentional, to group related functionality a bit.

**Comment by gwicke** [comment]
> Thinking now about double braces, wouldn't it be better to simply have them in URIs alone?

The intention behind the current code is to eventually make single quotes escape everywhere else as well, after a deprecation period & emitting warnings. This way, the behavior will be consistent between uri & other parts, and we can still use the escaped variants to construct URLs, for example in a JSON response containing a href, `Location:` headers, or `ETag`s.

**Comment by gwicke** [comment]
Hmmmmmmmm... After thinking about is some more, I'm wondering if we should use a syntax that doesn't require escaping in yaml. The main issue I see is that many characters are reserved there. Main candidates for a prefix would be `$` and `<`. Of these two, `$` (with a space) is somewhat reminiscent of a shell prompt, so perhaps invokes the right intuition?

Example:

``` yaml
request:
  uri: '{{options.host}}/{bar}'
  headers: 
    content-type: $ content-type
    cache-control: $ request.headers.cache-control
  body:
    foo: $ request.body.some.thing[3]
    bar: $ filter(request.body.arr, ['foo', 'bar'])
```

vs.

``` yaml
request:
  uri: '{{options.host}}/{bar}'
  headers: 
    content-type: '{{content-type}}'
    cache-control: '{{request.headers.cache-control}}'
  body:
    foo: '{{request.body.some.thing[3]}}'
    bar: '{{filter(request.body.arr, ['foo', 'bar'])}}'
```

**Comment by milimetric** [comment]
Hm.  I think `$ content-type` looks cool, but I also thinks coco looks cool and people almost flayed me for that.  So maybe `$.content-type` is more intuitive.

**Comment by gwicke** [comment]
Another option:

``` yaml
request:
  uri: '{{options.host}}/{bar}'
  headers: 
    content-type: <- content-type
    cache-control: <- request.headers.cache-control
  body:
    foo: <- request.body.some.thing[3]
```

**Comment by berndsi** [comment]
I agree it does look cool and concise. The `{{ }}` syntax looks more consistent, though, and I feel more confident to know when a variable part is terminated. The `uri` example for the other syntax options seems a bit odd to me in that it introduces the `+`. What does this mean? Why do you need it for `{+options.host}` but not `{bar}`? 
Why do you have `<-` or `$` for most of them but not for the `uri` example?

Ok, take this with a grain of salt since I only looked at this for a minute.

**Comment by gwicke** [comment]
@berndsi: Thanks, that's very helpful -- especially the bits that are strange to you.

I also changed the example from under you, sorry. The URI syntax is a subset of http://tools.ietf.org/html/rfc6570, in which `{+foo}` means "literal substitution without escaping", while `{foo}` will percent-escape the parameter. This patch adds support for `{{foo}}` as an alias of `{+foo}`, which is why I changed it back. The question was more about `'{{foo}}'` vs. `$ foo` in attributes that aren't multi-part string / URI templates.

**Comment by gwicke** [comment]
Merging in preparation for testing.

**Comment by berndsi** [comment]
The `<-` example still shows `{+options.host}` BTW.

**Comment by gwicke** [comment]
`$ fixed`

**Pull request merged**